src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs

```diff
@@ -0,0 +1,183 @@
+
+            return await GetReferenceItemsAsync(referenceParams, context, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<LSP.VSReferenceItem[]> GetReferenceItemsAsync(
+            LSP.ReferenceParams request,
+            SimpleFindUsagesContext context,
+            CancellationToken cancellationToken)
+        {
+            // Mapping each reference to its definition
+            var definitionMap = new Dictionary<DefinitionItem, List<SourceReferenceItem>>();
+            foreach (var reference in context.GetReferences())
+            {
+                if (!definitionMap.ContainsKey(reference.Definition))
+                {
+                    definitionMap.Add(reference.Definition, new List<SourceReferenceItem>());
+                }
+
+                definitionMap[reference.Definition].Add(reference);
+            }
+
+            // NOTE: Parts of FAR currently do not display correctly due to a bug in LSP.VSReferenceItem.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1088938/
+            using var _ = ArrayBuilder<LSP.VSReferenceItem>.GetInstance(out var referenceItems);
+
+            // Each reference is assigned its own unique id
+            var id = 0;
+            foreach (var (definition, references) in definitionMap)
+            {
+                // Creating the reference item that corresponds to the definition
+                var definitionItem = await GenerateReferenceItem(
+                    id, definitionId: id, definition.SourceSpans.FirstOrDefault(), context, definition.DisplayableProperties,
+                    cancellationToken, definition.GetClassifiedText()).ConfigureAwait(false);
+                referenceItems.Add(definitionItem);
+
+                var definitionId = id;
+                id++;
+
+                // Creating a reference item for each reference
+                foreach (var reference in references)
+                {
+                    var referenceItem = await GenerateReferenceItem(
+                        id, definitionId, reference.SourceSpan, context, reference.AdditionalProperties,
+                        symbolUsageInfo: reference.SymbolUsageInfo, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    referenceItems.Add(referenceItem);
+                    id++;
+                };
+            }
+
+            return referenceItems.ToArray();
+
+            static async Task<LSP.VSReferenceItem> GenerateReferenceItem(
+                int id,
+                int? definitionId,
+                DocumentSpan documentSpan,
+                SimpleFindUsagesContext context,
+                ImmutableDictionary<string, string> properties,
+                CancellationToken cancellationToken,
+                ClassifiedTextElement? definitionText = null,
+                SymbolUsageInfo? symbolUsageInfo = null)
```

Comment:
i would make these non-optiona.  then have the caller be explicit at they're intending to provide 'null' here as they're not relevant.

in general optional parameters for internal APIs are *highly* discouraged.